[css-animations] css/css-animations/animate-with-color-mix.html is a failure

graouts opened the pull request:
#### a08c3d1966f92e32a037a51a30b99567a536cd4b
<pre>
[css-animations] css/css-animations/animate-with-color-mix.html is a failure
<a href="https://bugs.webkit.org/show_bug.cgi?id=272083">https://bugs.webkit.org/show_bug.cgi?id=272083</a>

Reviewed by Anne van Kesteren and Darin Adler.

The `assert_oklab_color()` function in `css/support/color-testcommon.js` rejected our serialization
for `oklab()` colors because they included scientific notation. Indeed, this is incorrect, since CSS
numbers ought not be serialized using scientific notation, as the css-color spec indicates [0]:

    The component values are serialized in base 10; the L, a, b and C component values are serialized as &lt;number&gt;

And the CSSOM spec indicates the following for serialization of `&lt;number&gt;` [1]:

    A base-ten number using digits 0-9 (U+0030 to U+0039) in the shortest form possible, using “.” to
    separate decimals (if any), rounding the value if necessary to not produce more than 6 decimals,
    preceded by “-” (U+002D) if it is negative.
    NOTE: scientific notation is not used.

Credit to Anne van Kesteren for pointing out the relevant spec text here and find the root cause
of this issue. So all we needed to do was change the way we serialized `oklab()` and `oklch()` colors.

This change required some rebaselining or the addition of fuzzy comparison in the case of the WPT
test `css/css-color/parsing/color-valid-color-mix-function.html`.

[0] <a href="https://drafts.csswg.org/css-color-4/#serializing-oklab-oklch">https://drafts.csswg.org/css-color-4/#serializing-oklab-oklch</a>
[1] <a href="https://drafts.csswg.org/cssom/#serializing-css-values">https://drafts.csswg.org/cssom/#serializing-css-values</a>

* LayoutTests/css3/color/color-function-computed-style-expected.txt:
* LayoutTests/css3/color/color-function-computed-style.html:
* LayoutTests/css3/color/color-function-parsing-expected.txt:
* LayoutTests/css3/color/color-function-parsing.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-animations/animate-with-color-mix-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-color/parsing/color-computed-color-mix-function-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-color/parsing/color-computed-relative-color-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-color/parsing/color-mix-out-of-gamut-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-color/parsing/color-valid-color-mix-function-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-color/parsing/color-valid-color-mix-function.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-transitions/animations/animate-with-color-mix-expected.txt:
* LayoutTests/platform/ios/fast/forms/basic-inputs-expected.txt:
* LayoutTests/platform/ios/fast/forms/basic-textareas-expected.txt:
* LayoutTests/platform/ios/fast/forms/basic-textareas-quirks-expected.txt:
* LayoutTests/platform/ios/fast/forms/input-appearance-disabled-expected.txt:
* LayoutTests/platform/ios/fast/forms/input-disabled-color-expected.txt:
* LayoutTests/platform/ios/fast/forms/number/number-appearance-spinbutton-disabled-readonly-expected.txt:
* LayoutTests/platform/ios/fast/forms/placeholder-pseudo-style-expected.txt:
* LayoutTests/platform/ios/fast/forms/textarea-placeholder-pseudo-style-expected.txt:
* LayoutTests/platform/mac-ventura/fast/forms/basic-inputs-expected.txt:
* LayoutTests/platform/mac-ventura/fast/forms/input-disabled-color-expected.txt:
* LayoutTests/platform/mac-ventura/fast/forms/number/number-appearance-spinbutton-disabled-readonly-expected.txt:
* LayoutTests/platform/mac-ventura/fast/forms/placeholder-pseudo-style-expected.txt:
* LayoutTests/platform/mac/fast/forms/basic-inputs-expected.txt:
* LayoutTests/platform/mac/fast/forms/basic-textareas-expected.txt:
* LayoutTests/platform/mac/fast/forms/basic-textareas-quirks-expected.txt:
* LayoutTests/platform/mac/fast/forms/input-appearance-disabled-expected.txt:
* LayoutTests/platform/mac/fast/forms/input-disabled-color-expected.txt:
* LayoutTests/platform/mac/fast/forms/number/number-appearance-spinbutton-disabled-readonly-expected.txt:
* LayoutTests/platform/mac/fast/forms/placeholder-pseudo-style-expected.txt:
* LayoutTests/platform/mac/fast/forms/textarea-placeholder-pseudo-style-expected.txt:
* Source/WebCore/platform/graphics/ColorSerialization.cpp:
(WebCore::numericComponent):

Canonical link: <a href="https://commits.webkit.org/277256@main">https://commits.webkit.org/277256@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/820766431ac52e742f38e0a9ce663796502b4fd8

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/47112 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/26287 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/49751 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/49795 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/43161 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/49419 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/31565 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/23748 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/38377 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/47693 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/23770 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/40603 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/19684 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/21169 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/41754 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/5156 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/43498 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/42159 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/51671 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/22136 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/18512 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/45672 "Found 9 new test failures: fast/forms/basic-inputs.html, fast/forms/basic-textareas-quirks.html, fast/forms/basic-textareas.html, fast/forms/input-appearance-disabled.html, fast/forms/input-disabled-color.html, fast/forms/placeholder-pseudo-style.html, fast/forms/textarea-placeholder-pseudo-style.html, imported/w3c/web-platform-tests/css/css-color/parsing/color-computed-color-mix-function.html, imported/w3c/web-platform-tests/css/css-color/parsing/color-computed-relative-color.html (failure)") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/23413 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/40749 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/44677 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/24195 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/6623 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/23130 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

annevk reviewed (approved):
At some point this should probably share logic with `CSSPrimitiveValue::formatNumberValue`. Seems a bit suspect that serializes Nan and infinity differently.

graouts committed:
[css-animations] css/css-animations/animate-with-color-mix.html is a failure
https://bugs.webkit.org/show_bug.cgi?id=272083

Reviewed by Anne van Kesteren and Darin Adler.

The `assert_oklab_color()` function in `css/support/color-testcommon.js` rejected our serialization
for `oklab()` colors because they included scientific notation. Indeed, this is incorrect, since CSS
numbers ought not be serialized using scientific notation, as the css-color spec indicates [0]:

    The component values are serialized in base 10; the L, a, b and C component values are serialized as <number>

And the CSSOM spec indicates the following for serialization of `<number>` [1]:

    A base-ten number using digits 0-9 (U+0030 to U+0039) in the shortest form possible, using “.” to
    separate decimals (if any), rounding the value if necessary to not produce more than 6 decimals,
    preceded by “-” (U+002D) if it is negative.
    NOTE: scientific notation is not used.

Credit to Anne van Kesteren for pointing out the relevant spec text here and find the root cause
of this issue. So all we needed to do was change the way we serialized `oklab()` and `oklch()` colors.

This change required some rebaselining or the addition of fuzzy comparison in the case of the WPT
test `css/css-color/parsing/color-valid-color-mix-function.html`.

[0] https://drafts.csswg.org/css-color-4/#serializing-oklab-oklch
[1] https://drafts.csswg.org/cssom/#serializing-css-values

* LayoutTests/css3/color/color-function-computed-style-expected.txt:
* LayoutTests/css3/color/color-function-computed-style.html:
* LayoutTests/css3/color/color-function-parsing-expected.txt:
* LayoutTests/css3/color/color-function-parsing.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-animations/animate-with-color-mix-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-color/parsing/color-computed-color-mix-function-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-color/parsing/color-computed-relative-color-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-color/parsing/color-mix-out-of-gamut-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-color/parsing/color-valid-color-mix-function-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-color/parsing/color-valid-color-mix-function.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-transitions/animations/animate-with-color-mix-expected.txt:
* LayoutTests/platform/ios/fast/forms/basic-inputs-expected.txt:
* LayoutTests/platform/ios/fast/forms/basic-textareas-expected.txt:
* LayoutTests/platform/ios/fast/forms/basic-textareas-quirks-expected.txt:
* LayoutTests/platform/ios/fast/forms/input-appearance-disabled-expected.txt:
* LayoutTests/platform/ios/fast/forms/input-disabled-color-expected.txt:
* LayoutTests/platform/ios/fast/forms/number/number-appearance-spinbutton-disabled-readonly-expected.txt:
* LayoutTests/platform/ios/fast/forms/placeholder-pseudo-style-expected.txt:
* LayoutTests/platform/ios/fast/forms/textarea-placeholder-pseudo-style-expected.txt:
* LayoutTests/platform/mac-ventura/fast/forms/basic-inputs-expected.txt:
* LayoutTests/platform/mac-ventura/fast/forms/input-disabled-color-expected.txt:
* LayoutTests/platform/mac-ventura/fast/forms/number/number-appearance-spinbutton-disabled-readonly-expected.txt:
* LayoutTests/platform/mac-ventura/fast/forms/placeholder-pseudo-style-expected.txt:
* LayoutTests/platform/mac/fast/forms/basic-inputs-expected.txt:
* LayoutTests/platform/mac/fast/forms/basic-textareas-expected.txt:
* LayoutTests/platform/mac/fast/forms/basic-textareas-quirks-expected.txt:
* LayoutTests/platform/mac/fast/forms/input-appearance-disabled-expected.txt:
* LayoutTests/platform/mac/fast/forms/input-disabled-color-expected.txt:
* LayoutTests/platform/mac/fast/forms/number/number-appearance-spinbutton-disabled-readonly-expected.txt:
* LayoutTests/platform/mac/fast/forms/placeholder-pseudo-style-expected.txt:
* LayoutTests/platform/mac/fast/forms/textarea-placeholder-pseudo-style-expected.txt:
* Source/WebCore/platform/graphics/ColorSerialization.cpp:
(WebCore::numericComponent):

Canonical link: https://commits.webkit.org/277256@main
